[week8] API 문서 만들기, Docker image 만들기

caoyu-dev opened the pull request:
`Spring REST Docs` 를 활용하여 API 문서 작성 및 배운 내용을 정리하는 것이 이번 주 목표입니다. 😄

caoyu-dev committed:
[Add] asciidoctor, Spring Rest Docs, mariadb 설정
1. asciidoctor: 마크다운과 비슷하게 html 문서로 작성하기 위한 텍스트 작성 형식
2. Spring Rest Docs: API 문서 생성 자동화 도구
3. mariadb: MySql

caoyu-dev commented:
오전 중에 클래스 단위로 JavaDoc 작성하고 커밋하겠습니다 😎

hannut91 commented:
REST Docs를 만들고 여기에 스크린샷을 첨부해 주세요 😌

caoyu-dev commented:
<img width="1092" alt="스크린샷 2022-06-23 오후 11 46 18" src="https://user-images.githubusercontent.com/73353809/175328187-0e541af8-4ac6-44fc-a586-74702dab543c.png">

이렇게 올리는 걸까요? 역시 오전 업로드는 무리였어요 😆

caoyu-dev committed:
[Add] Test 의 한글 인코딩을 위해 UTF-8 인코딩 필터를 인터페이스화하여 추가

caoyu-dev committed:
[Add] Product 컨트롤러 테스트에 Rest docs 적용
1. test 코드 작성
2. snippet 파일 생성
3. adoc 파일 생성 + snippet 파일 import
4. html 파일로 변환
5. Rest docs 적용

caoyu-dev committed:
[Add] Product 관련 클래스에 Javadocs 작성

caoyu-dev commented:
목표는 내일 오전 중에 User 관련 클래스 정리 🤔

hannut91 commented:
좋은 API문서는 무엇일까요? 최근에 보신 문서중에 잘 만든 API 문서가 떠오르는게 있으신가요?

caoyu-dev commented:
> 좋은 API문서는 무엇일까요? 최근에 보신 문서중에 잘 만든 API 문서가 떠오르는게 있으신가요?

근 한달간은 회사 swagger 문서만 봤어요. 혹시 참고할 만한(혹은 추천할 만한) API 문서가 있으실까요? 
개인적으로 생각하는 **좋은 API 문서**란 **1) 반복되는 구조를 사용**하고, **2) 파라미터, 메서드 명 등의 단어 선택이 반복**되며, **3) 가독성이 좋고**, **4) 구조를 빠르게 파악 할 수 있는 문서**가 좋은 API 문서인 것 같습니다.

caoyu-dev committed:
[Add] User 관련 클래스에 Javadocs 작성

caoyu-dev committed:
[Add] User 관련 클래스에 Javadocs 작성
누락된 항목 커밋

caoyu-dev committed:
[Add] 인증, 인가 관련 클래스에 Javadocs 작성

caoyu-dev commented:
6/25 (토) 할 일
- [x] [Add] 인증, 인가 관련 클래스에 Javadocs 작성 - 누락된 항목 커밋
- [x] [Add] Error 관련 클래스에 Javadocs 작성
- [x] [Add] Test 에 Rest docs 적용
- [ ] 누락된 항목 생각하기 (일요일 보충)

hannut91 commented:
외부에 공개하는 API일수록 더 신경써서 만들겠죠. https://developers.kakao.com/docs/latest/ko/kakaologin/rest-api

caoyu-dev committed:
[Add] 인증, 인가 관련 클래스에 Javadocs 작성
누락된 항목 커밋

caoyu-dev committed:
[Add] Error 관련 클래스에 Javadocs 작성

caoyu-dev committed:
[Add] Test 에 Rest docs 적용

caoyu-dev committed:
[Add] index.adoc 에 register-user 추가
해당 파일로 작업 실행하여 HTML 파일 생성

caoyu-dev commented:
index.adoc 에 내용 추가하고, Gradle 로 `asciidoc` 작업을 실행하였습니다.
내용은 하기와 같습니다. 

<img width="1023" alt="스크린샷 2022-06-25 오후 11 28 57" src="https://user-images.githubusercontent.com/73353809/175778161-7a1492c2-99dd-4acc-a168-504957a35f4c.png">

----
코드숨 과정은 유익하고 좋은 시간이었습니다. 
시간의 여유가 조금만 더 따라줬다면 여러 가지를 실험해봤을 것 같아 약간의 아쉬움이 있지만요.

다음 주부터는 프로젝트로 뵙겠습니다. 그간 노고가 많으셨습니다.

caoyu-dev committed:
[Fix] DB 정보 변경

caoyu-dev committed:
[Fix] Java 버전 변경
build.gradle 에서 sourceCompatibility = '1.8' 을 '11' 로 변경
이유: 호환하는 버전이 달라 Usage of API document as @since 11+ 오류 발생

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -29,6 +32,14 @@ public AuthenticationService(UserRepository userRepository,
     }
 
+    /**
+     * 주어진 email 과 일치하는 User 의 password 를 확인 후 일치하면 암호화된 id 반환
```
암호화된 id는 조금 잘못해석 될 것 같고요. JWT정도로 써주면 될 것 같아요. JWT가 아무나 볼 수 있으니, 한 번 더 암호화를 하는 것도 있지만 대부분 그렇게 하지는 않구요. JWT는 서명이 된 값이죠. 그래서 서명한 키를 가지고 있지 않다면 조작할 수 없는 특징이 중요한 특징입니다.

caoyu-dev replied:
`user.getId()`를 `jwtUtil`에 있는 `encode` 메서드에 넣고 돌리기 때문에 id 를 암호화한다고 표현했는데 더 좋은 표현이 있었군요. 감사합니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -10,9 +10,15 @@
 @Service
 @Transactional
 public class ProductService {
+    /**
+     * Product 데이터 저장소
+     */
```
주석 위치가 잘못되었군요

caoyu-dev replied:
`ProductRepository` 상단에 올렸어야 했는데 `mapper` 위에 올라가 있네요. 😮! 잡아주셔서 감사합니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -53,6 +91,14 @@ public Product deleteProduct(Long id) {
 
+    /**
+     * 주어진 id 와 일치하는 Product 를 찾을 수 있으면 반환하고,
+     * 찾을 수 없다면 Product 를 찾을 수 없다는 예외 처리 반환
```
예외를 던진다 라고 하는 것은 어떨까요?

caoyu-dev replied:
예외 처리 반환 쓰면서 명료하지 못하다 느꼈는데, 윤석 님도 느끼셨나 보네요. 수정하겠습니다.

caoyu-dev replied:
`UserService` 쪽에도 동일한 어구가 있어 해당 주석도 수정했습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -63,12 +86,27 @@ public User updateUser(Long id, UserModificationData modificationData,
 
+    /**
+     * 주어진 id 와 일치하는 User 를 찾을 수 있으면 반환하고,
+     * 찾을 수 없다면 User 를 찾을 수 없다는 예외 처리 반환
```
여기도 마찬가지로 예외를 던진다라고 하면 좋을 것 같아요

caoyu-dev replied:
넵, 수정했습니다. 감사합니다. 😄

hannut91 commented on app/src/main/java/com/codesoom/assignment/domain/RoleRepository.java:
```diff
@@ -2,8 +2,22 @@
 
 import java.util.List;
 
+/**
+ * User 권한을 저장하고 반환
+ */
```
여기에 반환이라는 말은 적절하지 않은 것 같아요

caoyu-dev replied:
`User 권한을 저장` 으로 수정했습니다.

caoyu-dev committed:
[Fix] login 메서드 주석 수정
암호화된 id 반환 -> JWT 반환
JWT
1) JWT 는 아무나 볼 수 있어서 한 번 더 암호화 하는 경우도 있음
   -> 그러나 대부분 그렇게 하지는 않음
2) JWT 는 서명된 값이므로 서명한 키가 없으면 조작을 할 수 없는 것이 특징

caoyu-dev committed:
[Fix] RoleRepository 클래스에 대한 주석 수정

caoyu-dev committed:
[Fix] 메서드 주석 수정
1) ProductRepository 에 대한 주석 수정
2) 예외 처리 반환 -> 예외를 던진다 로 수정
